Python/errors.c

```diff
@@ -944,90 +944,167 @@ PyErr_NewExceptionWithDoc(const char *name, const char *doc,
+            Py_DECREF(moduleName);
+            if (PyFile_WriteString(".", file) < 0) {
+                return -1;
+            }
+        }
+        else {
+            Py_DECREF(moduleName);
         }
     }
     if (className == NULL) {
-        if (PyFile_WriteString("<unknown>", f) < 0)
-            goto done;
+        if (PyFile_WriteString("<unknown>", file) < 0) {
+            return -1;
+        }
     }
     else {
-        if (PyFile_WriteString(className, f) < 0)
-            goto done;
+        if (PyFile_WriteString(className, file) < 0) {
+            return -1;
+        }
     }
 
-    if (v && v != Py_None) {
-        if (PyFile_WriteString(": ", f) < 0)
-            goto done;
-        if (PyFile_WriteObject(v, f, Py_PRINT_RAW) < 0) {
+    if (exc_value && exc_value != Py_None) {
+        if (PyFile_WriteString(": ", file) < 0) {
+            return -1;
+        }
+        if (PyFile_WriteObject(exc_value, file, Py_PRINT_RAW) < 0) {
             PyErr_Clear();
-            if (PyFile_WriteString("<exception str() failed>", f) < 0) {
-                goto done;
+            if (PyFile_WriteString("<exception str() failed>", file) < 0) {
+                return -1;
             }
         }
     }
-    if (PyFile_WriteString("\n", f) < 0)
-        goto done;
+    if (PyFile_WriteString("\n", file) < 0) {
+        return -1;
+    }
+    return 0;
+}
 
-done:
-    Py_XDECREF(moduleName);
-    Py_XDECREF(t);
-    Py_XDECREF(v);
-    Py_XDECREF(tb);
+
+PyObject*
+_PyErr_WriteUnraisable(PyObject *exc_type, PyObject *exc_value,
+                       PyObject *exc_tb, PyObject *obj)
+{
+    if (err_write_unraisable(exc_type, exc_value, exc_tb, obj, 0) < 0) {
+        return NULL;
+    }
+    Py_RETURN_NONE;
+}
+
+
+static void
+unraisable_hook_failed(void)
+{
+    PyObject *exc_type, *exc_value, *exc_tb;
+    PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
+
+    err_write_unraisable(exc_type, exc_value, exc_tb, NULL, 1);
+
+    Py_XDECREF(exc_type);
+    Py_XDECREF(exc_value);
+    Py_XDECREF(exc_tb);
+    PyErr_Clear(); /* Just in case */
+}
+
+
+
+/* Call when an exception has occurred but there is no way for Python
+   to handle it.  Examples: exception in __del__ or during GC.
+
+   Call sys.unraisablehook(exc_type, exc_value, exc_tb, obj). */
+void
+PyErr_WriteUnraisable(PyObject *obj)
+{
+    PyObject *exc_type, *exc_value, *exc_tb;
+
+    PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
+
+    _Py_IDENTIFIER(unraisablehook);
+    PyObject *hook = _PySys_GetObjectId(&PyId_unraisablehook);
+    if (hook != NULL && hook != Py_None) {
+        PyObject *args[4];
+        args[0] = (exc_type ? exc_type : Py_None);
+        args[1] = (exc_value ? exc_value : Py_None);
+        args[2] = (exc_tb ? exc_tb : Py_None);
+        args[3] = (obj ? obj : Py_None);
+        PyObject *res = _PyObject_FastCall(hook, args, Py_ARRAY_LENGTH(args));
+        Py_XDECREF(res);
+        if (res == NULL) {
+            unraisable_hook_failed();
```

Comment:
I suggest to just fall back to the standard handler called below.
```c
if (hook != NULL && hook != Py_None) {
    PyErr_Fetch(&exc_type, &exc_value, &exc_tb);
    ... // call the hook
    Py_XDECREF(exc_type);
    Py_XDECREF(exc_value);
    Py_XDECREF(exc_tb);
    if (res != NULL) {
        goto done;
    }
    obj = hook;
}
write_unraisable(...);
done:
```

Reply:
I inlined unraisable_hook_failed() to make the fallback more explicit.

Reply:
Did you forget to push your changes?

Reply:
I didn't write exactly what you proposed, but what I wrote is basically the same without goto. I dislike goto and "obj = hook" here. I prefer a more explicit fallback.

Reply:
Well. I modified my PR to write an explicit fallback as you asked me for :-)